File: viewer/src/main/scala/farjs/viewer/quickview/QuickViewPanel.scala
package farjs.viewer.quickview

import farjs.filelist.{FileListData, FileListState}
import farjs.filelist.api.FileListItem
import farjs.filelist.stack.{WithStack, WithStacks}
import farjs.filelist.theme.FileListTheme
import farjs.ui._
import farjs.ui.border._
import scommons.nodejs.path
import scommons.react._
import scommons.react.blessed._

object QuickViewPanel extends FunctionComponent[Unit] {

  private[quickview] var doubleBorderComp: ReactClass = DoubleBorder
  private[quickview] var horizontalLineComp: ReactClass = HorizontalLine
  private[quickview] var textLineComp: ReactClass = TextLine
  private[quickview] var quickViewDirComp: ReactClass = QuickViewDir
  private[quickview] var quickViewFileComp: UiComponent[QuickViewFileProps] = QuickViewFile

  protected def render(compProps: Props): ReactElement = {
    val stacks = WithStacks.useStacks()
    val panelStack = WithStack.useStack()
    val width = panelStack.width
    val height = panelStack.height
    
    val theme = FileListTheme.useTheme().fileList
    val stack =
      if (!panelStack.isRight) stacks.right.stack
      else stacks.left.stack
    val stackItem = stack.peek[FileListState]()

    val maybeCurrData = stackItem.getData().toOption.flatMap {
      case FileListData(dispatch, actions, state) =>
        FileListState.currentItem(state).toOption.map {
          case i if i == FileListItem.up => (dispatch, actions, state, FileListItem.currDir)
          case i => (dispatch, actions, state, i)
        }
    }

    <.box(^.rbStyle := theme.regularItem)(
      <(doubleBorderComp)(^.plain := DoubleBorderProps(width, height, theme.regularItem))(),
      <(horizontalLineComp)(^.plain := HorizontalLineProps(
        left = 0,
        top = height - 4,
        length = width,
        lineCh = SingleChars.horizontal,
        style = theme.regularItem,
        startCh = DoubleChars.leftSingle,
        endCh = DoubleChars.rightSingle
      ))(),
      <(textLineComp)(^.plain := TextLineProps(
        align = TextAlign.center,
        left = 1,
        top = 0,
        width = width - 2,
        text = "Quick view",
        style = theme.regularItem,
        focused = panelStack.stack.isActive
      ))(),

      maybeCurrData.map { case (dispatch, actions, state, currItem) =>
        <.>()(
          if (currItem.isDir) {
            <(quickViewDirComp)(^.plain := QuickViewDirProps(
              dispatch = dispatch,
              actions = actions,
              state = state,
              stack = panelStack.stack,
              width = width,
              currItem = currItem
            ))()
          }
          else {
            val filePath = path.join(state.currDir.path, currItem.name)
            <.box(
              ^.rbLeft := 1,
              ^.rbTop := 1,
              ^.rbWidth := width - 2,
              ^.rbHeight := height - 5,
              ^.rbStyle := theme.regularItem
            )(
              <(quickViewFileComp())(
                ^.key := filePath,
                ^.plain := QuickViewFileProps(
                  dispatch = dispatch,
                  panelStack = panelStack,
                  filePath = filePath,
                  size = currItem.size
                )
              )()
            )
          },

          <.text(
            ^.rbWidth := width - 2,
            ^.rbHeight := 2,
            ^.rbLeft := 1,
            ^.rbTop := height - 3,
            ^.rbStyle := theme.regularItem,
            ^.content := currItem.name
          )()
        )
      }
    )
  }
}

File: viewer/quickview/QuickViewPanel.mjs
/**
 * @import { ThemeEffects } from "@farjs/ui/theme/Theme.mjs"
 */
import path from "path";
import React, { useRef } from "react";
import TextLine from "@farjs/ui/TextLine.mjs";
import TextAlign from "@farjs/ui/TextAlign.mjs";
import SingleChars from "@farjs/ui/border/SingleChars.mjs";
import DoubleChars from "@farjs/ui/border/DoubleChars.mjs";
import DoubleBorder from "@farjs/ui/border/DoubleBorder.mjs";
import HorizontalLine from "@farjs/ui/border/HorizontalLine.mjs";
import WithStacks from "@farjs/filelist/stack/WithStacks.mjs";
import WithStack from "@farjs/filelist/stack/WithStack.mjs";
import FileListTheme from "@farjs/filelist/theme/FileListTheme.mjs";
import FileListItem from "@farjs/filelist/api/FileListItem.mjs";
import FileListState from "@farjs/filelist/FileListState.mjs";
import QuickViewDir from "./QuickViewDir.mjs";
import QuickViewFile from "./QuickViewFile.mjs";

const h = React.createElement;

const QuickViewPanel = () => {
  const {
    doubleBorderComp,
    horizontalLineComp,
    textLineComp,
    quickViewDirComp,
    quickViewFileComp,
  } = QuickViewPanel;

  const stacks = WithStacks.useStacks();
  const panelStack = WithStack.useStack();
  const inputRef = useRef(panelStack.panelInput);
  const width = panelStack.width;
  const height = panelStack.height;

  const theme = FileListTheme.useTheme().fileList;
  /** @type {ThemeEffects} */
  const style = theme.regularItem;

  const stack = !panelStack.isRight ? stacks.right.stack : stacks.left.stack;
  const maybeData = [stack.peek().getData()]
    .filter((_) => _ !== undefined)
    .flatMap((data) => {
      return [FileListState.currentItem(data.state)]
        .filter((_) => _ !== undefined)
        .map((i) => {
          return {
            data,
            currItem: i === FileListItem.up ? FileListItem.currDir : i,
          };
        });
    });

  return h(
    "box",
    { style },
    h(doubleBorderComp, {
      width: width,
      height: height,
      style,
    }),
    h(horizontalLineComp, {
      left: 0,
      top: height - 4,
      length: width,
      lineCh: SingleChars.horizontal,
      style,
      startCh: DoubleChars.leftSingle,
      endCh: DoubleChars.rightSingle,
    }),
    h(textLineComp, {
      align: TextAlign.center,
      left: 1,
      top: 0,
      width: width - 2,
      text: "Quick view",
      style,
      focused: panelStack.stack.isActive,
    }),

    maybeData.length > 0
      ? (([{ data, currItem }]) => {
          const filePath = path.join(data.state.currDir.path, currItem.name);
          return h(
            React.Fragment,
            null,
            currItem.isDir
              ? h(quickViewDirComp, {
                  dispatch: data.dispatch,
                  actions: data.actions,
                  state: data.state,
                  stack: panelStack.stack,
                  width,
                  currItem,
                })
              : h(
                  "box",
                  {
                    left: 1,
                    top: 1,
                    width: width - 2,
                    height: height - 5,
                    style,
                  },
                  h(quickViewFileComp, {
                    key: filePath,
                    dispatch: data.dispatch,
                    inputRef: inputRef,
                    isRight: panelStack.isRight,
                    filePath,
                    size: currItem.size,
                  })
                ),

            h("text", {
              width: width - 2,
              height: 2,
              left: 1,
              top: height - 3,
              style,
              content: currItem.name,
            })
          );
        })(maybeData)
      : null
  );
};

QuickViewPanel.displayName = "QuickViewPanel";
QuickViewPanel.doubleBorderComp = DoubleBorder;
QuickViewPanel.horizontalLineComp = HorizontalLine;
QuickViewPanel.textLineComp = TextLine;
QuickViewPanel.quickViewDirComp = QuickViewDir;
QuickViewPanel.quickViewFileComp = QuickViewFile;

export default QuickViewPanel;
